File: 백준/Gold/2493. 탑/탑.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();

        int n = Integer.parseInt(br.readLine());
        StringTokenizer st = new StringTokenizer(br.readLine());

        Stack<int[]> stack = new Stack<>();

        for (int i = 1; i <= n; i++) {
            int height = Integer.parseInt(st.nextToken());

            while (!stack.isEmpty() && stack.peek()[1] < height) {
                stack.pop();
            }

            if (stack.isEmpty()) {
                sb.append("0 ");
            } else {
                sb.append(stack.peek()[0]).append(" ");
            }

            stack.push(new int[]{i, height});
        }

        System.out.print(sb);
    }
}

File: 백준/Gold/2493. 탑/탑.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*

fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {
    val n = readLine().toInt()
    val st = StringTokenizer(readLine(), " ")
    val sb = StringBuilder()
    val stack = Stack<Pair<Int, Int>>()
    repeat(n) { idx ->
        val num = st.nextToken().toInt()
        if (stack.empty()) {
            sb.append("0 ")
            stack.push(Pair(idx, num))
        } else {
            while (true) {
                if (stack.empty()) {
                    sb.append("0 ")
                    stack.push(Pair(idx, num))
                    break
                }
                if (stack.peek().second > num) {
                    sb.append("${stack.peek().first + 1} ")
                    stack.push(Pair(idx, num))
                    break
                } else {
                    stack.pop()
                }
            }
        }
    }
    println(sb)
}
